src/index.rs

```diff
@@ -268,3 +268,337 @@ fn is_hidden(entry: &DirEntry) -> bool {
+            );
+
+            let mut actual = ResourceIndex::build(path.clone())
+                .expect("Could not build index");
+
+            assert_eq!(actual.collisions.len(), 0);
+            assert_eq!(actual.size(), 2);
+
+            file.set_permissions(Permissions::from_mode(0o222))
+                .expect("Should be fine");
+
+            // let mut name_from = path.clone();
+            // name_from.push(FILE_NAME_2);
+            // let mut name_to = path.clone();
+            // path.clone().push(FILE_NAME_3);
+            // std::fs::rename(name_from, name_to);
+
+            let update = actual
+                .update()
+                .expect("Should update index correctly");
+
+            assert_eq!(actual.collisions.len(), 0);
+            assert_eq!(actual.size(), 2);
+            assert_eq!(update.deleted.len(), 0);
+            assert_eq!(update.added.len(), 0);
+        })
+    }
+
+    // error cases
+
+    #[test]
+    fn should_not_index_empty_file() {
+        run_test_and_clean_up(|path| {
+            create_file_at(path.clone(), Some(0), None);
+            let actual = ResourceIndex::build(path.clone())
+                .expect("Could not generate index");
+
+            assert_eq!(actual.root, path.clone());
+            assert_eq!(actual.path2meta.len(), 0);
+            assert_eq!(actual.ids.len(), 0);
+            assert_eq!(actual.collisions.len(), 0);
+        })
+    }
+
+    #[test]
+    fn should_not_index_hidden_file() {
+        run_test_and_clean_up(|path| {
+            create_file_at(path.clone(), Some(FILE_SIZE_1), Some(".hidden"));
+            let actual = ResourceIndex::build(path.clone())
+                .expect("Could not generate index");
+
+            assert_eq!(actual.root, path.clone());
+            assert_eq!(actual.path2meta.len(), 0);
+            assert_eq!(actual.ids.len(), 0);
+            assert_eq!(actual.collisions.len(), 0);
+        })
+    }
+
+    #[test]
+    fn should_not_index_1_empty_directory() {
+        run_test_and_clean_up(|path| {
+            create_dir_at(path.clone());
+
+            let actual = ResourceIndex::build(path.clone())
+                .expect("Could not build index");
+
+            assert_eq!(actual.root, path.clone());
+            assert_eq!(actual.path2meta.len(), 0);
+            assert_eq!(actual.ids.len(), 0);
+            assert_eq!(actual.collisions.len(), 0);
+        })
+    }
+
+    #[test]
+    fn should_fail_when_indexing_file_without_read_rights() {
+        run_test_and_clean_up(|path| {
+            let file = create_file_at(path.clone(), Some(1), None);
+            file.set_permissions(Permissions::from_mode(0o222))
+                .expect("Should be fine");
+
+            let actual =
+                std::panic::catch_unwind(|| ResourceIndex::build(path.clone()));
+            assert!(actual.is_err());
+        })
+    }
+
+    #[test]
+    fn discover_paths_should_not_walk_on_invalid_path() {
```

Comment:
Actually, it seems that we need to provide user app with clear error in this case.

Reply:
Sure, let's keep this in mind for now and address in a followup issue as it probably involves code change as well.